### PR TITLE
Financial updates: LCOE calc and grouping logic

### LINKED ISSUE #175: LCOE should only include cost for technologies involved in producing electricity by default
# LCOE should only include cost for technologies involved in producing electricity by default

While working on #165, we found that some parts of the current financial stackup don't match common practices.
Specifically, LCOE should only include costs from electricity-producing techs.
We can use the definitions in `supported_models.py` for this.
There may be future complications when multiple technologies that don't directly create electricity are used in the hybrid plant; that is out of scope for this issue.

Also, sort of on a separate note but we uncovered this at the same time so we might as well fix it simultaneously: the behavior for how financial groups exist is not clear when you specify a financial group for one tech but not another.
We should clean this up so that if you don't specify a group the framework assumes everything is in the same group to avoid confusion.

### PR DESCRIPTION
**Pull request by johnjasa**
# Financial updates: LCOE calc and grouping logic

Addressing some small financial modeling updates that came about when working on #165.
- LCOE and LCOH metrics are now computed based on lists optionally given by users
- Minor fix for grouping of financial calcs (edge case), added test

## PR Checklist

- [x] `CHANGELOG.md` has been updated to describe the changes made in this PR
- [x] Documentation
  - [x] Docstrings are up-to-date
  - [x] Related `docs/` files are up-to-date, or added when necessary
  - [x] Documentation has been rebuilt successfully
  - [x] Examples have been updated
- [x] Tests pass (If not, and this is expected, please elaborate in the tests section)
- [x] PR description thoroughly describes the new feature, bug fix, etc.

## Related issues

Resolves #175

### CODE REVIEW

File: h2integrate/core/h2integrate_model.py [hunk 1]
```diff
@@ -275,9 +275,16 @@ def create_financial_model(self):
...
-        if not financial_groups:
-            financial_groups["1"] = self.technology_config["technologies"]
+        # Find technologies not assigned to any financial group and add them to group "1"
+        all_grouped_techs = set()
+        for group in financial_groups.values():
+            all_grouped_techs.update(group.keys())
+
+        for tech_name, tech_config in self.technology_config["technologies"].items():
+            if tech_name not in all_grouped_techs:
+                if 1 not in financial_groups:
```
[the remark this answers is not in the document]
Comment (johnjasa) [reply h2integrate/core/h2integrate_model.py:285]: Yes, that is correct

File: h2integrate/core/finances.py [hunk 2]
```diff
@@ -204,6 +207,11 @@ def compute(self, inputs, outputs):
         )
         pf.set_params("cash onhand", self.plant_config["finance_parameters"]["cash_onhand_months"])
 
+        last_tech_that_produces_electricity = None
+        for tech in self.tech_config:
+            if tech in electricity_producing_techs:
+                last_tech_that_produces_electricity = tech
+
         # --------------------------------- Add capital and fixed items to ProFAST --------------
         for tech in self.tech_config:
```
[the remark this answers is not in the document]
Comment (johnjasa) [reply h2integrate/core/finances.py:237]: Great point, we talked through this together and I changed the code introduced in this PR so users can directly specify which technologies are included in the metrics.

File: h2integrate/core/h2integrate_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (johnjasa) [reply h2integrate/core/h2integrate_model.py:285]: Do you want that behavior to change? Maybe have a `default` group instead of `1`? Or something else?

[the remark this answers is not in the document]
Comment (jaredthomas68) [reply h2integrate/core/h2integrate_model.py:300]: I agree with Kaitlin here. I think the default group name should be a name that a user is not likely to set intentionally, like 'default'.

File: tests/h2integrate/test_all_examples.py [hunk 3]
```diff
@@ -126,9 +126,6 @@ def test_wind_h2_opt_example(subtests):
     with subtests.test("Check LCOH"):
         assert model.prob.get_val("financials_group_1.LCOH")[0] < 4.64
```
Comment (jaredthomas68) [inline tests/h2integrate/test_all_examples.py:127]: I know this is not part of this PR per se, but It think it would be better to change the 4.64 on the right to the initial LCOH before optimization so that if the models or assumptions change, the test remains valid and useful.

**Review by jaredthomas68 — CHANGES_REQUESTED** [review]
Thanks for this John, looks good generally. I left a couple of small comments, the only one I see as blocking is one Kaitlin had already pointed out.

**Review by kbrunik — CHANGES_REQUESTED** [review]

File: tests/h2integrate/test_all_examples.py (hunk 3 above)
Comment (johnjasa) [reply tests/h2integrate/test_all_examples.py]: Got it, good call, I modified the case and test!

**Comment by johnjasa** [comment]
Thanks for your reviews, @kbrunik and @jaredthomas68! Ready for re-reviews.

I've modified the default group so it's now `financial_group_default` and removed the automatic grouping based on the last tech producing electricity. Instead, users must now define exactly what technologies they want included for any LCO metric or _all_ technologies will be included.

File: docs/user_guide/how_to_set_up_an_analysis.ipynb [hunk 4]
```diff
@@ -171,6 +171,26 @@
         "If the list is length 4, then the third entry in the list is what's passing passed via a transporter of the type defined in the fourth entry.\n",
         "If the list is length 3, then the third entry in the list is what is connected directly between the technologies.\n",
         "\n",
+        "Within the `plant_config` file, you can define which technologies are included in the levelized cost of commodities calculations.\n",
+        "This is done by defining the list of those technologies in the `finance_parameters` section like this:\n",
+        "\n",
+        "```\n",
+        "finance_parameters:\n",
```
[the remark this answers is not in the document]
Comment (johnjasa) [reply docs/user_guide/how_to_set_up_an_analysis.ipynb:178]: Fixed!
